tools/perfscale-load-generator/executor/executor.go

```diff
@@ -0,0 +1,115 @@
+
+	"kubevirt.io/kubevirt/tools/perfscale-load-generator/config"
+)
+
+const scenarioLabel = "virt-load-generator-scenario"
+
+// Executor contains the information required to execute a job
+type Executor struct {
+	Start  time.Time
+	End    time.Time
+	Config config.ConfigSpec
+	// Limits the number of workers to QPS and Burst
+	limiter *rate.Limiter
+}
+
+// NewExecutorList Returns a executor
+func NewExecutorList(conf config.ConfigSpec) *Executor {
+	return &Executor{
+		Config:  conf,
+		limiter: rate.NewLimiter(rate.Limit(conf.Scenario.QPS), conf.Scenario.Burst),
+	}
+}
+
+func (e Executor) Run() {
+	e.Start = time.Now().UTC()
+	var namespace string
+
+	for i := 1; i <= e.Config.Scenario.IterationCount; i++ {
+		log.Log.V(2).Infof("Creating %d VMIs from iteration %d", e.Config.Scenario.VMICount, i)
+
+		if e.Config.Scenario.NamespacedIterations || i == 1 {
+			namespace = fmt.Sprintf("%s-%d", e.Config.Scenario.Namespace, i)
+			if err := CreateNamespaces(e.Config.ClientSet, namespace, e.Config.Scenario.UUID); err != nil {
+				log.Log.V(2).Error(err.Error())
+				continue
+			}
+			log.Log.V(2).Infof("Created namespace %s", namespace)
+		}
+
+		e.createVMI(namespace, i)
+
+		if e.Config.Scenario.IterationVMIWait {
+			log.Log.V(2).Infof("Waiting %s for VMIs in namespace %v to be in the Running phase", e.Config.Scenario.MaxWaitTimeout, namespace)
+			if err := WaitForRunningVMIs(e.Config.ClientSet, namespace, e.Config.Scenario.UUID, e.Config.Scenario.VMICount, e.Config.Scenario.MaxWaitTimeout); err != nil {
+				log.Log.V(2).Errorf("Failed to create VMIs: %v", err)
+			} else {
+				log.Log.V(2).Infof("%d VMIs were sucessfully created in namespace %v in %v", e.Config.Scenario.VMICount, namespace, time.Since(e.Start))
+			}
+		}
+
+		DeleteVMIs(e.Config.ClientSet, namespace, e.Config.Scenario.UUID, e.limiter)
+
+		if e.Config.Scenario.IterationWaitForDeletion {
+			log.Log.V(2).Infof("Waiting %s for VMIs in namespace %v to be deleted", e.Config.Scenario.MaxWaitTimeout, namespace)
+			if err := WaitForDeleteVMIs(e.Config.ClientSet, namespace, e.Config.Scenario.UUID); err != nil {
+				log.Log.V(2).Errorf("Failed to delete VMIs: %v", err)
+			} else {
+				log.Log.V(2).Infof("All VMIs were deleted")
+			}
+		}
+
+		if e.Config.Scenario.IterationInterval > 0 {
+			log.Log.V(2).Infof("Sleeping for %v between interations", e.Config.Scenario.IterationInterval)
+			time.Sleep(e.Config.Scenario.IterationInterval)
+		}
+
+		if e.Config.Scenario.IterationCleanup {
+			log.Log.V(2).Infof("Clean up all created namespaces")
+			WaitForDeleteVMIs(e.Config.ClientSet, namespace, e.Config.Scenario.UUID)
+			if err := CleanupNamespaces(e.Config.ClientSet, e.Config.Scenario.UUID); err != nil {
+				log.Log.V(2).Errorf("Error cleaning up namespaces: %v", err)
+			}
+		}
+	}
+	e.End = time.Now().UTC()
+	log.Log.V(2).Infof("Scenario Startup Time: %v", e.Start)
+	log.Log.V(2).Infof("Scenario End Time: %v", e.End)
+}
+
+func (e Executor) createVMI(ns string, i int) {
```

Comment:
```suggestion
func (e Executor) createVMIs(ns string, i int) {
```

Also somewhere you mention parallel creation, would it be a good idea to provide `MaxParallelRequests` and then create VMIs in goroutines to simulate a parallel load instead if a sequential one?

Reply:
Right, maybe I should remove the parallel word. 

Since the creation is not waiting for the VMI creation, it is only sending the request to create the object, I do not see too much benefit from gorotine. It is a short life command.
We control the QPS with the `limiter`.

Reply:
If you don't fire them in parallel you don't truly simulate concurrent traffic like from many different clients.
This might not be the goal, then fine. But on the API side it can make a difference if all the requests come after another or at the same time (like the load on our webhooks etc.).